### PR TITLE
Typed reducers

### PR DESCRIPTION
**Pull request by nglazov**
I suggest to put actions and reducers in modules folder, it will be more comfortable to work with it (when you are changing reducer you are changing actions and when they are in one folder it's more fast)

Also i've added typed actions in reducer and when you are writing reducers case you know which payload is there.

I moved one component inside of container, it allow to decrease loc (you don't need to define typings twice) and used typeof and ResolveThunks to avoid extra typings

### CODE REVIEW

File: cvat-ui/src/containers/register-page/register-page.tsx [hunk 1]
```diff
@@ -1,38 +1,80 @@
...
+    onRegister: typeof registerAsync;
 }
 
-function mapStateToProps(state: CombinedState): StateToProps {
-    return {
-        fetching: state.auth.fetching,
+type Props = StateToProps & ResolveThunks<DispatchToProps> & RouteComponentProps;
+
+function RegisterPageContainer(props: Props) {
+    const sizes = {
```
Comment (mmarkelov) [inline cvat-ui/src/containers/register-page/register-page.tsx]: It's better to move out of function

File: cvat-ui/src/containers/register-page/register-page.tsx [hunk 2]
```diff
@@ -1,38 +1,80 @@
...
+    } = props;
+
+    return (
+        <Row type='flex' justify='center' align='middle'>
+            <Col {...sizes}>
+                <Title level={2}> Create an account </Title>
+                <RegisterForm
+                    fetching={fetching}
+                    onSubmit={(registerData: RegisterData): void => {
+                        onRegister(
```
Comment (mmarkelov) [inline cvat-ui/src/containers/register-page/register-page.tsx]: I suppose using object here will be better

Comment (nglazov) [reply cvat-ui/src/containers/register-page/register-page.tsx]: It was just moved code without any changes

File: cvat-ui/src/modules/auth/authReducer.ts [hunk 3]
```diff
@@ -53,7 +52,7 @@ export default (state = defaultState, action: AnyAction): AuthState => {
             return {
                 ...state,
                 fetching: true,
-                user: action.payload.user,
+                user: undefined,
```
[the remark this answers is not in the document]
Comment (nglazov) [reply cvat-ui/src/modules/auth/authReducer.ts]: I've used value by types: i had empty object and user of {} is undefined

[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/modules/auth/authReducer.ts]: It was a bug

File: cvat-ui/package.json [hunk 4]
```diff
@@ -14,12 +18,15 @@
     "@babel/preset-env": "^7.6.0",
     "@babel/preset-react": "^7.0.0",
     "@babel/preset-typescript": "^7.6.0",
-    "@typescript-eslint/eslint-plugin": "^2.10.0",
+    "@typescript-eslint/eslint-plugin": "^2.19.2",
+    "@typescript-eslint/experimental-utils": "^2.19.2",
```
Comment (bsekachev) [inline cvat-ui/package.json]: I can't see where you use this package

**Comment by bsekachev** [comment]
@nglazov 

There is a merge conflict

File: cvat-ui/package.json (hunk 4 above)
Comment (nglazov) [reply cvat-ui/package.json]: Yes, i think it was added, when i tried to run eslint local...
It looks like "babel": "^6.23.0" may be removed to, is not it?

Comment (bsekachev) [reply cvat-ui/package.json]: @nglazov

Yep, it looks like babel is also not needed.

File: cvat-ui/src/utils/redux.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+export interface ActionWithPayload<T, P> extends Action<T> {
+    payload: P;
+}
+
+export function createAction<T extends string>(type: T): Action<T>;
+export function createAction<T extends string, P>(type: T, payload: P): ActionWithPayload<T, P>;
+export function createAction<T extends string, P>(type: T, payload?: P) {
+    return typeof payload === 'undefined' ? { type } : { type, payload };
+}
+
```
Comment (bsekachev) [inline cvat-ui/src/utils/redux.ts]: I am not very get on well with typescript, and I am trying to sort it out. 

They are overloads, aren't they?
I [read](https://howtodoinjava.com/typescript/function-overloading/), that all overloads must have the same return type

Comment (bsekachev) [reply cvat-ui/src/utils/redux.ts]: Why do not extend type redux::Action by adding a payload property that has a generic type?

Something like that:
```ts
interface Action<T extends string, P>  implements  ReduxAction<T> {
    type: T,
    payload: P,
}

export function createAction<T extends string>(type: T): Action<T, undefined>;
export function createAction<T extends string, P>(type: T, payload: P): Action<T, P>;
export function createAction<T extends string, P>(type: T, payload?: P) {
    return typeof payload === 'undefined' ? { type } : { type, payload };
}

```

Comment (bsekachev) [reply cvat-ui/src/utils/redux.ts]: It's more question, than offer.

Comment (bsekachev) [reply cvat-ui/src/utils/redux.ts]: Oh, actually you do almost the same, but nevertheless return different types.

File: cvat-ui/src/cvat-store.ts [hunk 6]
```diff
@@ -21,9 +22,10 @@ const middlewares = [
 let store: Store | null = null;
 
 export default function createCVATStore(createRootReducer: () => Reducer): void {
+    // todo enable devtools only in dev mode
```
Comment (bsekachev) [inline cvat-ui/src/cvat-store.ts]: So, for now does it work in production?
Do you happen to know how it affects performance?

File: cvat-ui/src/utils/redux.ts (hunk 5 above)
Comment (bsekachev) [reply cvat-ui/src/utils/redux.ts]: Okay, I will investigate about

File: cvat-ui/src/cvat-store.ts (hunk 6 above)
Comment (nglazov) [reply cvat-ui/src/cvat-store.ts]: It's easier to remove todo:)

File: cvat-ui/package.json (hunk 4 above)
Comment (nglazov) [reply cvat-ui/package.json]: Fixed

**Pull request merged**